mindmeld/models/embedder_models.py

```diff
@@ -0,0 +1,127 @@
+# limitations under the License.
+
+"""
+This module contains the embedder model class.
+"""
+from abc import ABC, abstractmethod
+import pickle
+import logging
+import os
+
+from .. import path
+from .helpers import register_embedder
+
+logger = logging.getLogger(__name__)
+
+try:
+    from sentence_transformers import SentenceTransformer
+
+    register_bert = True
+except ImportError:
+    logger.warning("Must install the extra [bert] to use the built in embbedder.")
+    register_bert = False
+
+
+class Embedder(ABC):
+    """
+    Base class for embedder model
+    """
+
+    def __init__(self, app_path, embedder_type, **kwargs):
+        """Initializes an embedder.
+        """
+        self.cache_path = path.get_embedder_cache_file_path(app_path, embedder_type)
+
+        folder = os.path.dirname(self.cache_path)
+        if not os.path.isdir(folder):
+            os.makedirs(folder)
+
+        if not os.path.exists(self.cache_path) or os.path.getsize(self.cache_path) == 0:
+            self.cache = {}
+        else:
+            with open(self.cache_path, "rb") as fp:
+                self.cache = pickle.load(fp)
+
+        self.model = self.load(**kwargs)
+
+    @abstractmethod
+    def load(self, **kwargs):
+        """Loads the embedder model
+
+        Returns:
+            The model object.
+        """
+        pass
+
+    @abstractmethod
+    def encode(self, text_list):
+        """
+        Args:
+            text_list (list): A list of text strings for which to generate the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays of the embeddings.
+        """
+        pass
+
+    def clear_cache(self):
+        """Deletes the cache file.
+        """
+        if os.path.exists(self.cache_path):
+            os.remove(self.cache_path)
+
+    def get_encodings(self, text_list):
+        """Fetches the encoded values from the cache, or generates them.
+        Args:
+            text_list (list): A list of text strings for which to get the embeddings.
+
+        Returns:
+            (list): A list of numpy arrays with the embeddings.
+        """
+        encoded = [self.cache.get(text, None) for text in text_list]
+        cache_miss_indices = [i for i, vec in enumerate(encoded) if vec is None]
+        text_to_encode = [text_list[i] for i in cache_miss_indices]
+        model_encoded_text = self.encode(text_to_encode)
+
+        for i, v in enumerate(cache_miss_indices):
+            encoded[v] = model_encoded_text[i]
+            self.cache[text_to_encode[i]] = model_encoded_text[i]
+        return encoded
+
+    def dump(self):
+        """Dumps the cache to disk.
+        """
+        with open(self.cache_path, "wb") as fp:
+            pickle.dump(self.cache, fp)
+
+
+class BertEmbedder(Embedder):
+    """
+    Encoder class for bert models as described here: https://github.com/UKPLab/sentence-transformers
+    """
+
+    def load(self, **kwargs):
```

Comment:
maybe we should add a way for the developers to use their own custom BERT model?

```
# Use BERT for mapping tokens to embeddings
word_embedding_model = models.Transformer('path/to/your/BERT/model')

# Apply mean pooling to get one fixed sized sentence vector
pooling_model = models.Pooling(word_embedding_model.get_word_embedding_dimension(),
                               pooling_mode_mean_tokens=True,
                               pooling_mode_cls_token=False,
                               pooling_mode_max_tokens=False)

model = SentenceTransformer(modules=[word_embedding_model, pooling_model])

Reply:
Developers can do this by implementing their own embedder inheriting from the abstract class:

from mindmeld.models import Embedder, register_embedder

class MyEmbedder(Embedder):
    def load(self, **kwargs):
        # return the loaded model
    def encode(self, text_list):
        # encodes each query in the list 

register_embedder('my_embedder_name', MyEmbedder)

I think it's cleaner to have them use this functionality rather than supporting too much variation in the built in version. But let me know if you think otherwise.